marimo/_runtime/executor.py

```diff
@@ -0,0 +1,389 @@
+
+
+def build_ref_predicate(
+    glbls: dict[str, Any],
+) -> Callable[[Name, VariableData], bool]:
+    """
+    All declared variables are tied together under the graph of required_refs.
+    Strict execution gets the minimum graph of definitions for execution.
+    Certain definitions, like lambdas, functions, and classes contain an
+    executable body and require their `required_refs` to be scope (included in
+    this graph). This function determines if a potential reference should be
+    included in the graph based on its computed type. Consider:
+
+    >>> def foo():
+    ...     return bar()
+
+    here `foo` is a function with `bar` as a reference in the execution body,
+    so if `foo` is a reference, both `bar` and `foo` should be included in the
+    graph, otherwise we'll get a NameError on `bar` if `foo` is called.
+    Compare that to:
+
+    >>> x = foo()
+
+    if `x` is the only reference, should `foo` be included in the graph? It
+    depends on the context, so we defer to the type of `x` which has already
+    been computed at this point. If `x` is a known 'primitive' type, and thus
+    does not have an executable body, we can exclude `foo` from the graph.
+    However, `foo` may return a object or another function, which in turn may
+    have references; so if x doesn't match the very low bar 'primitive', its
+    `required_refs` are included in the graph.
+
+    NB: lambdas, as anonymous functions, do not have a name to refer to them-
+    so visitor injects the dummy variable `_lambda` into the `required_refs` to
+    denote their presence.
+    """
+
+    def check_ref(ref: Name) -> bool:
+        return ref in glbls and (
+            inspect.isfunction(glbls[ref])
+            or inspect.ismodule(glbls[ref])
+            or inspect.isclass(glbls[ref])
+            or callable(glbls[ref])
+        )
+
+    def only_scoped_refs(ref: Name, data: VariableData) -> bool:
+        # Weakref instances should be disassociated from related references,
+        # as should other "primitives" as they are results and hopefully not
+        # hiding some scoped reference. Other common types could be added here,
+        # like numpy arrays that are not dtype=object, etc.. that are known not
+        # to be dependent on the functions that created them.
+
+        # This errs on the side of including too much, but that's a better user
+        # experience than not having definitions available.
+        return (
+            ref in glbls
+            and not isinstance(glbls[ref], PRIMITIVES)
+            and (
+                "_lambda" in data.required_refs
+                or any(map(check_ref, data.required_refs | {ref}))
+            )
+        )
+
+    return only_scoped_refs
+
+
+def is_instance_by_name(obj: object, name: str) -> bool:
+    if not (hasattr(obj, "__module__") and hasattr(obj, "__class__")):
+        return False
+    obj_name = f"{obj.__module__}.{obj.__class__.__name__}"
+    return obj_name == name
```

Comment:
Fancy — are you doing this so the module doesn't have to be installed?

Reply:
That's the intent; not certain if it works right now

I think the last change I want to make is to test/get this working since it seemed like it wasn't catching some things. That will probably have to wait for tomorrow evening.

Reply:
Actually, wasn't that bad. pushed this too